text/nodejs/0014-yarn-berry.md

```diff
@@ -0,0 +1,280 @@
+  desire to leverage Plug'n'Play and should not need to run `yarn install`
+  whether or not the `yarn.lock` has changed between builds.
+
+
+The buildpack will store the Yarn cache in a layer and configure Yarn to
+reference this layer, ideally via the `YARN_CACHE_FOLDER` environment variable.
+For Berry projects using `node_modules`, the behaviour will largely mimic the
+existing Yarn Install buildpack, with accommodations for the CLI changes.
+
+The buildpack will generate a filesystem-based SBoM as `yarn-install` does currently.
+
+### Layer Reuse 
+
+The Yarn Install buildpack currently uses a checksum of the output from `yarn
+config list` and the contents of `yarn.lock` to decide whether to reuse package layers.
+`yarn config list` has been removed in Berry, but the new `yarn info` command, which
+provides information about the project's direct and transitive dependencies,
+should work well as a replacement.
+
+### `--frozen-lockfile`
+
+The existing Yarn Install buildpack passes the `--frozen-lockfile` argument to
+every `yarn install` call, which will not update the `yarn.lock` file, and will
+fail if the `yarn.lock` file needs to be updated. As such, developers must run
+`yarn install` locally in order to update their app's `yarn.lock` file.
+`--frozen-lockfile` has been deprecated in favour of `--immutable`, which this
+buildpack will utilize.
+
+
+### Offline builds
+
+Previously, the `yarn-install` buildpack relied on `yarn-offline-mirror` along
+with the `--offline` flag to facilitate offline builds. This workflow is no
+longer supported and is indeed unnecessary since `.yarn/cache` stores archives
+of all packages now anyway.
+
+### Running Scripts
+
+Berry has [changed the way pre- & post- scripts are
+invoked](https://yarnpkg.com/getting-started/migration#explicitly-call-the-pre-and-post-scripts).
+
+The build logic of `yarn-start` will need to be modified to accommodate this
+change.
+
+### `yarn node` vs `node` 
+
+All `node` commands run via the shell should be run with `yarn node` instead. The default
+start command in the Yarn Start buildpack may need to be modified.
+
+## Rationale and Alternatives
+
+- Include logic for Yarn Berry workflows in the existing `yarn-install`
+  buildpack.
+
+  Berry includes breaking changes to the `yarn` CLI used by the buildpack to
+  install dependencies. The changes to CLI options are such that at the very
+  least, the buildpack would need to decide which ones to use based on the type
+  of application being built. Trying to incorporate both workflows will
+  increase the scope of responsibility of the buildpack, likely increasing its
+  complexity at the expense of maintainability. Classic Yarn is now in
+  maintenance mode and, though it is still heavily used now (due to the
+  relatively slow adoption of Berry), its unclear how long it will be
+  supported. For this reason, it is even more prudent to keep the logic
+  separate to facilitate the eventual transition.
```

Comment:
I think it might be reasonable to still get Berry support in the existing buildpack if we just take some effort to keep the "Classic" and Berry parts separate. If we did something like move all the existing code in the buildpack into a `classic` package and then created a new `berry` package we can keep them logically separated. We could then just have a small amount of code at the top-level that decides with codepath to invoke.

Reply:
I hadn't considered separate packages, that's an interesting suggestion. It would remove a lot of overhead in the event that we drop support for Classic altogether.